### PR TITLE
Add support for contiguous data format

### PR DESCRIPTION
**Pull request by cjh1**
This add a need file format version (5) to support contiguous sector layout.

### CODE REVIEW

**Comment by ercius** [comment]
What's the status of this pull request? We can test it tomorrow if there is a singularity image to pull.

**Comment by cjh1** [comment]
@ercius This should be ready to go, it would be great if you can test it. I will push a new image.

**Comment by cjh1** [comment]
`openchemistry/stempy:latest` is now built from this branch. Let me know if you run into any problems.

File: stempy/reader.cpp [hunk 1]
```diff
@@ -278,9 +278,22 @@ int extractSector(const std::string& fileName)
...
 }
 
-SectorStreamReader::SectorStreamReader(const vector<string>& files)
-  : m_files(files)
+SectorStreamReader::SectorStreamReader(const vector<string>& files,
+                                       uint8_t version)
+  : m_files(files), m_version(version)
 {
+  // Validate version
+  switch (this->m_version) {
```
Comment (cryos) [inline stempy/reader.cpp:286]: When using m_ for member variables it would be preferable to avoid using this-> as well

File: stempy/reader.cpp [hunk 2]
```diff
@@ -278,9 +278,22 @@ int extractSector(const std::string& fileName)
...
+    case 4:
+    case 5:
+      break;
+    default:
+      std::ostringstream ss;
+      ss << "Unsupported version: ";
+      ss << this->m_version;
```
Comment (cryos) [inline stempy/reader.cpp:293]: As above

**Comment by ercius** [comment]
Thanks! I'll test it today.

**Comment by ercius** [comment]
I just tried this on the test data 2020.04.29/data_scan5*.data but on the processing computer. Its not reading the files correctly. When trying to do electron counting I get these results for the fitting:
```
****Statistics for calculating electron thresholds****
number of samples:1200
min sample:-32768
max sample:32256
mean:115.434
variance:354272
std dev:595.208
number of bins:65024
x-ray threshold n sigma:175
background threshold n sigma:4
optimized mean:9.40402
optimized std dev:7.70378
background threshold:40.2191
xray threshold:104277
Segmentation fault (core dumped)
```
The min and max are outside uint16 values. This is with the backend='thread' option.

Without that option I get this error:
```python
Traceback (most recent call last):
  File "/home/daquser/temp_pae/electron_count.py", line 46, in <module>
    background_threshold_n_sigma=threshold)
  File "/usr/local/lib/python3.7/site-packages/stempy/image/__init__.py", line 324, in electron_count
    blocks.append(next(reader))
  File "/usr/local/lib/python3.7/site-packages/stempy/io/__init__.py", line 19, in __next__
    b = self.read()
  File "/usr/local/lib/python3.7/site-packages/stempy/io/__init__.py", line 31, in read
    b = super(ReaderMixin, self).read()
ValueError: Unexpected EOF while processing stream.
```
Can you test that your container works when you pull stempy:latest using:
`singularity pull docker://openchemistry/stempy:latest`

Im calling the electron counting with these lines:
```python
sReader = stio.reader(iFiles,stio.FileVersion.VERSION5, backend='thread')
ee = stim.electron_count(sReader,dark0,number_of_samples=1200,
                                            verbose=True,threshold_num_blocks=20,
                                            xray_threshold_n_sigma=175,
                                            background_threshold_n_sigma=threshold)
```

**Comment by cjh1** [comment]
@ercius Can you test removing the `backend='thread'`?

**Comment by cjh1** [comment]
I ran the test myself, seems to work in serial mode, so I will do some digging ....

**Comment by ercius** [comment]
I just tested without the backend='thread' also. That works. I made a mistake in my earlier test where I removed the fileversions and the backend keyword. 

So, the only issue is with the threaded reader and the contiguous format.

**Comment by ercius** [comment]
This is working now. Im using the stempy-threaded singularity image with the threaded option. So, this can be added to master.

**Comment by cjh1** [comment]
@ercius Excellent, thanks for testing.

**Comment by cjh1** [comment]
@ercius This is now merge, I have updated in the images ( including the images on cori ).

**Comment by ercius** [comment]
Awesome. Thanks for getting that all set up. I'll test Cori when I get a
chance.

**Comment by ercius** [comment]
Tested on Cori and it's working well. Threading on Haswell nodes are only 25% slower than our local analysis PC.

**Comment by cjh1** [comment]
That is great news. Was that from a burst buffer? Interactive or batch?

**Comment by cjh1** [comment]
@ercius I just saw the email, thanks.

**Pull request merged**